src/model/Stroke.cpp

```diff
@@ -168,40 +168,32 @@ void Stroke::move(double dx, double dy) {
     this->sizeCalculated = false;
 }
 
-void Stroke::rotate(double x0, double y0, double xo, double yo, double th) {
+void Stroke::rotate(double x0, double y0, double th) {
+    cairo_matrix_t rotMatrix;
+    cairo_matrix_init_identity(&rotMatrix);
+    cairo_matrix_translate(&rotMatrix, x0, y0);
+    cairo_matrix_rotate(&rotMatrix, th);
+    cairo_matrix_translate(&rotMatrix, -x0, -y0);
+
     for (auto&& p: points) {
-        p.x -= x0;  // move to origin
-        p.y -= y0;
-        double offset = 0.7;  // __DBL_EPSILON__;
-        p.x -= xo - offset;   // center to origin
-        p.y -= yo - offset;
-
-        double x1 = p.x * cos(th) - p.y * sin(th);
-        double y1 = p.y * cos(th) + p.x * sin(th);
-        p.x = x1;
-        p.y = y1;
-
-        p.x += x0;  // restore the position
-        p.y += y0;
-
-        p.x += xo - offset;  // center it
-        p.y += yo - offset;
+        cairo_matrix_transform_point(&rotMatrix, &p.x, &p.y);
     }
     // Width and Height will likely be changed after this operation
     calcSize();
 }
 
-void Stroke::scale(double x0, double y0, double fx, double fy) {
-    double fz = sqrt(fx * fy);
+void Stroke::scale(double x0, double y0, double fx, double fy, double rotation, bool doScale) {
+    double fz = (doScale) ? sqrt(fx * fy) : 1;
```

Comment:
Just some notes, for me the functions needs to be evaluated for a refactor:
 - doesn't fx*fy == 1 imply, that doScale is false?
 - it look some kind of odd, that there is a rotation used to scale, this should be irrelevant, when previous translations are applied, or the element childs are positioned relative to to the element, isn't it?
 - the thickness `fz` should be calculated from the angle it has (Law of cosines), not just the pythagorean theorem. Currently the scaling is only correct for perfectly diagonal lines.
 - also this function looks more like a general translation function, maybe renaming it / creating a new would be more clear.

Reply:
To scale the line width and pressure of points with the geometric mean of the scaling factors was not my idea, but makes some sense to me.

- In the special case where fx*fy==1 then the line width and pressure of points is left unmodified. The boolean doScale is about whether line width and pressure should be potentially modified after a selection has been resized.
- The rotation angle for this generalized scale function is only used for the Undo Handler (see ScaleUndoAction.cpp:46). The Undo Handler does not remember how the selection was rotated, when the scaling was done, it just remembers what it should undo/redo with the elements of the selection. Note that this function does scale in a rotated coordinate system. So instead of scaling coordinates by fx in direction of the usual x-axis and by fy in direction of the usual y-axis, it does scaling with respect to the rotated x- and y-axis.
- I think the thickness fz should be the same for the whole stroke, not depend on the angle of the single line segments. Otherwise, if you want to do it with "correct scaling", you get ugly strokes. This is already visible if you scale a rectangle (with round line caps) only in x-direction (or y-direction). See the following example:
![grafik](https://user-images.githubusercontent.com/40485433/84573174-20357100-ad9f-11ea-8a4c-8504cb6b5037.png)

- If you prefer, I can remove the rotation and change the undo / redo action. However the rotation function also uses a rotation center, so it is also not just standard rotation around the origin.